proposals/blittable.md

```diff
@@ -0,0 +1,161 @@
+
+``` c#
+struct SimplePoint
+{
+    public int X;
+    public int Y;
+}
+
+blittable struct Data
+{
+    // Error: blittable struct cannot contain field of type SimplePoint which is not blittable
+    SimplePoint Point;
+}
+```
+
+The language will also support the ability to constrain generic type parameters to be `blittable` types.  
+
+``` C#
+void M<T>(T p) where T : blittable struct
+{
+    ...
+}
+
+M<Point>(); // Ok
+M<User>();  // Error: Type User does not satisfy the blittable constraint.
+```
+
+One of the primary motivations for `blittable` structs is ease of interop.  As such it's imperative that such a type have it's field ordered in a sequential, or explicit, layout.  An auto layout of fields makes it impossible to reliably interop the data.  
+
+The default today is for a sequential layout so this doesn't represent a substantial change.  However the compiler will make it illegal to mark such types as having an auto layout.  
+
+``` c#
+// Error: A blittable struct may not be marked with an auto layout
+[StructLayout(LayoutKind.Auto)]
+blittable struct LayoutExample 
+{
+    ...
+}
+```
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+The primary drawback of this feature is that it serves a small number of developers: typically low level library authors or frameworks.  Hence it's spending precious language time for a small number of developers. 
+
+Yet these frameworks are are often the basis for the majority of .NET applications out there.  Hence performance / correctness wins at this level can have a riple effect on the .NET ecosystem.  This makes the feature worth considering even with the limited audience. 
+
+There will also likely be a small transition period after this is released where core libraries move to adopt it.  Types like [System.Runtime.InteropServices.ComTypes.FILETIME](https://msdn.microsoft.com/en-us/library/system.runtime.interopservices.comtypes.filetime(v=vs.110).aspx) are common in interop code.  Until it is marked as `blittable` in source though, developers won't be able to depend on it in their libraries.  
+
+## Alternatives
+[alternatives]: #alternatives
+
+There are a couple of alternatives to consider:
+
+- The status quo:  The feature is not justified on its own merits and developers continue to use the implicit opt in behavior.
+- Generic constraint only: The blittable keyword is used on generic constraints only.  This allows for developers to author efficient helper libraries.  But the types involved lack any declarative support and hence are fragile across distributed development. 
+
+
+## Unresolved questions
+[unresolved]: #unresolved-questions
+
+- blittable vs. unmanaged.  The F# language has a very [similar feature](https://docs.microsoft.com/en-us/dotnet/articles/fsharp/language-reference/generics/constraints) which uses the keyword unmanaged. The blittable name comes from the use in Midori.  May want to look to precedence here and use unmanaged instead. 
```

Comment:
"blittable type" has an existing definition in CLR interop already: https://docs.microsoft.com/en-us/dotnet/framework/interop/blittable-and-non-blittable-types . The existing definition is different from your definition, e.g. `char` and `bool` are not blittable in the existing definition. It may be a good idea to call it something else to avoid confusion.

Reply:
It's probably worthwhile to keep this as blittable and remove `char` and `bool` from the definition.

That being said, `bool` and `char` are special in that they are considered "sometimes" blittable and, last I checked, the (CoreCLR) marshaller will still blit `bool` and `char` when the appropriate conditions are met (just like it will pin and directly marshal a `string` when the right conditions are met).

Reply:
The existing definition also have "The following complex types are also blittable types: One-dimensional arrays of blittable types". What would you do about that?

Reply:
I think we are going to go with the term "unmanaged type". It is already in use in the C# spec and F#.